app/controllers/generators_controller.rb

```diff
@@ -43,57 +43,64 @@ def create
-    params[:generator]["noid"] = new_noid.to_s.rjust(6, "0")
-    params[:generator]["last_date"] = last_date
-
-    respond_to do |format|
-      if @generator.update(generator_params)
-        project = params[:generator]["project"]
-        class_code = params[:generator]["class_code"]
-        prefix = params[:generator]["prefix"]
-        suffix = params[:generator]["suffix"]
-        picture_code = params[:generator]["picture_code"]
-        date_taken = params[:generator]["date_taken"]
-        noid = params[:generator]["noid"]
-        year = Time.zone.now.strftime("%Y")
-        month = Time.zone.now.strftime("%m")
-        if project.present? || (["Bulletin", "Mosley Photographs"].include? @generator.name)
-          case @generator.name
-          when "General"
-            message = "NOID: " + "#{project}Z#{year}#{month}#{noid}"
-          when "Oral Histories"
-            message = "NOID: " + "#{project}JZ#{year}#{month}#{noid}"
-          when "Templana (Complex)"
-            message = "NOID: " + "#{project}X#{class_code}Z#{year}#{month}#{noid}"
-          when "Bulletin"
-            message = "NOID: " + "#{prefix}Z#{year}#{month}#{noid}#{suffix}"
-          when "Mosley Photographs"
-            message = "NOID: " + "BPA001X#{picture_code.upcase}#{date_taken}#{noid}"
+    @generator.with_lock do
+      g = Generator.uncached { Generator.find(@generator.id) }
+      version = @generator.lock_version
+      last_date = Time.now.strftime("%m")
+      if g.last_date == last_date
+        new_noid = g.noid.to_i + 1
+      else
+        new_noid = 1
+      end
+      params[:generator]["noid"] = new_noid.to_s.rjust(6, "0")
+      params[:generator]["last_date"] = last_date
+
+      respond_to do |format|
+        if @generator.update(generator_params)
+          project = params[:generator]["project"]
+          class_code = params[:generator]["class_code"]
+          prefix = params[:generator]["prefix"]
+          suffix = params[:generator]["suffix"]
+          picture_code = params[:generator]["picture_code"]
+          date_taken = params[:generator]["date_taken"]
+          noid = params[:generator]["noid"]
+          year = Time.zone.now.strftime("%Y")
+          month = Time.zone.now.strftime("%m")
+          if project.present? || (["Bulletin", "Mosley Photographs"].include? @generator.name)
+            case @generator.name
+            when "General"
+              message = "NOID: " + "#{project}Z#{year}#{month}#{noid}"
+            when "Oral Histories"
+              message = "NOID: " + "#{project}JZ#{year}#{month}#{noid}"
+            when "Templana (Complex)"
+              message = "NOID: " + "#{project}X#{class_code}Z#{year}#{month}#{noid}"
+            when "Bulletin"
+              message = "NOID: " + "#{prefix}Z#{year}#{month}#{noid}#{suffix}"
+            when "Mosley Photographs"
+              message = "NOID: " + "BPA001X#{picture_code.upcase}#{date_taken}#{noid}"
+            else
+              return
```

Comment:
What happens when you get this step?

Reply:
This determines the message displayed back to the user, i.e. the "noid"
Different projects have different formats. These case conditions just set the appropriately formatted string that has the noid value for that generator and any other values as set in the project.

Reply:
I understand that part.  What I don't understand is that this is a control action method that supposedly responds to the user with something as shown on these liines if it gets to them https://github.com/tulibraries/noid_generators/blob/16a74bdfa3d836eb6b18c684c06a7978f6514647/app/controllers/generators_controller.rb#L93-L98.   But with this `return` here it would never head to a render step which is usually how an action method is supposed to terminate (i.e. with something back to the user).